src/coreclr/System.Private.CoreLib/src/System/ValueType.cs

```diff
@@ -42,7 +44,9 @@ public override bool Equals([NotNullWhen(true)] object? obj)
             // if there are no GC references in this object we can avoid reflection
```

Comment:
This check above:
```
            if (thatType != thisType)
```
can be changed to:
```
            if (this.GetType() != obj.GetType())
```
And materializing of the types can be moved to be after the `CanCompareBits`.

`this.GetType() != obj.GetType()` should treated as an intrinsics by the JIT, so this will make the fast memcmp path faster.